tidy3d/components/source.py

```diff
@@ -296,22 +296,35 @@ def amp_time(self, time: float) -> complex:
 
 
 class CustomSourceTime(Pulse):
-    """Custom source time dependence, real or complex valued.
+    """Custom source time dependence, real or complex valued, modulated at a central frequency.
 
     Note
     ----
     The source time dependence is linearly interpolated to the simulation time steps.
     To ensure that this interpolation does not introduce artifacts, it is necessary
     to use a sampling rate that is sufficiently fast relative to the simulation time step.
+    The source should also ideally start at zero and ramp up smoothly.
+    The first and last values of the envelope will be used for times that are out of range
+    of the provided data.
 
     Example
     -------
     >>> cst = td.CustomSourceTime.from_values(freq0=1, fwidth=0.1,
     >>>     values=np.linspace(0, 9, 10), dt=0.1)
     """
 
+    offset: float = pydantic.Field(
+        0.0,
+        title="Offset",
+        description="Time delay of the maximum value of the "
```

Comment:
so now the source amplitude time dependences is `f(t) = real[ source_time_dataset(t) * exp(i 2 pi freq0 t) ] `? And the offset just shifts `t` -> `t - offset`? the description should probably be modified if so.

Reply:
guess the offset is actually dimensionless so you need to incorporate fwidth but same idea?

Reply:
Actually the offset only shifts the envelope but not the carrier wave. The carrier wave can be shifted by `source_time.phase`.